[feat][monitor] Add rkt-query-inspector-monitor - feat/rtk query monitor

FaberVitale opened the pull request:
#  PR: feat/rtk query monitor

## Description

This PR adds a new monitor that displays [RTK query](https://redux-toolkit.js.org/rtk-query/overview) queries and mutations for [Redux DevTools](https://github.com/gaearon/redux-devtools).

The idea for this project has started as a discussion at `#redux` channel and has evolved in a discussion in a [redux-toolkit issue](https://github.com/reduxjs/redux-toolkit/issues/1097).

More informations are available at [rtk-query-inspector-monitor readme](https://github.com/FaberVitale/redux-devtools/blob/feat/rtk-query-monitor/packages/redux-devtools-rtk-query-inspector-monitor/README.md)

## Demo

* demo: https://rtk-query-monitor-demo.netlify.app/

## Previews

### Dock monitor

![RTK Query inspector monitor demo](https://raw.githubusercontent.com/FaberVitale/redux-devtools/feat/rtk-query-monitor/packages/redux-devtools-rtk-query-inspector-monitor/monitor-demo.gif)

### App - default theme inverted
<img width="793" alt="Screenshot 2021-06-20 at 11 23 26" src="https://user-images.githubusercontent.com/16583125/122668789-161c2e80-d1ba-11eb-92ad-b12a00adfa59.png">

### App - default theme
<img width="791" alt="Screenshot 2021-06-20 at 11 23 07" src="https://user-images.githubusercontent.com/16583125/122669219-92b00c80-d1bc-11eb-9607-c47ce5bf71cc.png">
---

## Notable repo changes

* added `@redux-devtools/rtk-query-inspector-monitor` to monorepo

* added `packages/redux-devtools/examples/rtk-query-polling`

* integrated `@redux-devtools/rtk-query-inspector-monitor` to `@redux-devtools/app`

* added `@types/react` as monorepo devDep

## PR Status - TODO

### `rtk-query-inspector-monitor`

`rtk-query-inspector-monitor` has almost reached all the features planned, [see TODO section for more details](https://github.com/FaberVitale/redux-devtools/blob/feat/rtk-query-monitor/packages/redux-devtools-rtk-query-inspector-monitor/README.md#todo), but feedback and new ideas are welcome.

### Integrations

* Verify extension integration, It wasn't clear how to start a dev server and what 'port is using.

* Verify monitor on browsers that runs on windows: ugly scrollbars? other issues?

* Fix linting errors that have emerged in `redux-devtools-serialize`

* Other?

Methuselah96 commented:
Wow, thanks for working on this! I haven't gotten a chance to look at the code yet, but what do you think about naming it just `rtk-query-monitor`? There could be some confusion about whether this is part of the `inspector-monitor`, but from what I can tell this is entirely separate from the `inspector-monitor`, so it might be more clear if we drop the term `inspector` from the package name. Plus, shorter names are always a nice to have when we can get away with it. :)

FaberVitale commented:
Hi @Methuselah96,

Thanks for all the work you have put in on the redux-devtools :)

> [...] but what do you think about naming it just rtk-query-monitor? 

Yeah, It wasn't the best naming choice in hindsight :)

I'll change it to `rtk-query-monitor`  👍 .


### Extension

I could use some help with the extension:

I wanted to test the integration with the extension but it isn't clear  if `start` spins up a dev server.
I tried running `npx serve  <build-dir>` but It has a reference error.
Would you kindly give me some guidance on this?

Methuselah96 commented:
The way that I've been testing the extension is by running `build:extension` and then loading it into Chrome using "Load unpacked" (here are some details on how to do that: https://developer.chrome.com/docs/extensions/mv3/getstarted/#manifest). It may be that the same process works with `start`, but I haven't tried it yet.

Note that there was a major overhaul of `@redux-devtools/app` (https://github.com/zalmoxisus/remotedev-app/pull/30) that hadn't been integrated into the extension yet when the previous maintainer went away and I got on the scene. I have started working on integrating the updated libraries into the extension in https://github.com/reduxjs/redux-devtools/pull/747, but it is still very much a work in progress, so be aware that there may be some rough edges. I am actively working on finishing the extension integration.

FaberVitale commented:
Changes

1. renamed package to `@redux-devtools/rtk-query-monitor`
2. improved UI of sort order control.
3. added timings section to api stats.
4. fixed (hopefully) CI errors.
5. added fallback message in case there are no rtk-query apis.
6. small internal refactors.

-  [updated demo](https://app.netlify.com/sites/rtk-query-monitor-demo/overview)

FaberVitale commented:
Changes:

* **(new)** actions tab: displays pending, rejected or fulfilled actions of selected query.
* improved responsiveness of query list.
* improved responsiveness of tab list.
* **(new)** displays median of active queries.
* **(new)** displays fetching duration of selected query.
* improved performance

- [Updated Demo](https://rtk-query-monitor-demo.netlify.app/)

phryneas commented:
I think by now this looks great, and while there might always be something to be improved, I would love to see if we can get this out soon, so just users can already profit from it. And getting user feedback will also only happen when we get this out - at least as a beta of the extension (I remember this was at least possible in the Firefox extension store, not sure for Chrome).

@FaberVitale There is one more feature that might be a great addition, but I'm not even sure if it would/should be part of this PR: Can you add a checkbox to the default action monitor to show/hide RTKQ-related actions? I think one of the biggest critiques towards RTKQ in terms of debugging was that it dispatches a lot of actions, and makes non-RTKQ-code more difficult to debug. So having a switch for just hiding those automated actions would be super-cool.

@Methuselah96 I am not really aware of the state of your migration on this repo - are we currently in a state where we could just release a new version of the devtools or is there any necessary work required that would have to be done before that?

FaberVitale commented:
@phryneas  in terms of features I believe the monitor is in good shape and it could be released on npm. 

### Things left to do

#### Blocking

 - I'd like to improve the [demo dependencies](https://github.com/FaberVitale/redux-devtools/blob/feat/rtk-query-monitor/packages/redux-devtools-rtk-query-monitor/demo/package.json#L13).
  It seems that you have written a possible solution [here](https://github.com/reduxjs/redux-toolkit/blob/master/package.json#L38).

#### Nice to have (non blocking)

- Download button(s).

- Improve UI of API stats. 
  I do not think that a tree view is the best way to render those info.

- other?

---

### Checkbox to hide RTKQ actions

> PR: Can you add a checkbox to the default action monitor to show/hide RTKQ-related actions? 

Yes and I would not mind doing it, but it is outside the scope of this PR: 
it's on another monitor.

Your idea is related to https://github.com/reduxjs/redux-devtools/issues/717 and I think I could implement them both.
In terms of UX/UI the vscode search checkboxes would fit our requirements.
<img width="136" alt="Screenshot 2021-07-18 at 15 33 58" src="https://user-images.githubusercontent.com/16583125/126069119-58b83977-e41f-4f02-bae8-51961bd3df6b.png">

What do you think?

Methuselah96 commented:
@phryneas Definitely looking forward to getting this released! I haven't quite finished all the work necessary to release the 3.x extension with the new UI and it would be a little bit of work to get this working in 2.x, I have three things left to do before we can release a 3.0 beta:

1. Finish converting the extension to TypeScript (I'm 80% done with this, should be able to finish this week). This has also allowed me to learn the overall architecture of the extension.
2. Touch up on things that I overlooked when integrating the new UI into the extension. This includes:
  a. The extension specific buttons that aren't included in `@redux-devtools/app`.
  b. The persistence middleware that used to be in `@redux-devtools/app`, but was removed in the UI rewrite. I need to look into why it was removed and if it needs to be replaced.
3. Do more through testing to make sure I didn't break anything when integrating the new UI.

I'm thinking it will take me 2-4 weeks to finish up the work I listed above. I also need to look over this PR to make sure everything's in order before we merge it, but it looks great so far!

markerikson commented:
Awesome! Let us know if there's any help we can provide!

phryneas commented:
@FaberVitale I think that would go well with #717 - and yup, I think those UI elements would do it :)

@Methuselah96 that sounds great :) As for the "thorough testing" part, maybe we can also just put out an alpha or beta and see what feedback comes in :)

FaberVitale commented:
### Changes

- improved ui of api section
- added toggleable regex search **(new)**
- mutations are displayed in the `queryList` (left section) **(new)**

Updated demo: https://rtk-query-monitor-demo.netlify.app/

phryneas commented:
This is looking very good :)
One thing I noticed is that especially playing with mutations, the "Tally" is maybe a bit useless - as mutations are immediately removed when a new mutation is queued, the number will always be 1 here - and maybe confuse people a bit.  

But it might just need another title... "Current cache contents summed up" or something?

Same goes for the timings, especially mutation timings: since they only sum up what is currently in the cache, their use is a bit limited. But I guess tracking historical data is too much work, right?

FaberVitale commented:
> One thing I noticed is that especially playing with mutations, the "Tally" is maybe a bit useless - as mutations are immediately removed when a new mutation is queued, the number will always be 1 here - and maybe confuse people a bit.

I agree

> Same goes for the timings, especially mutation timings: since they only sum up what is currently in the cache, their use is a bit limited. But I guess tracking historical data is too much work, right?

It is doable but instead of looking for `cacheKey` in the latest user state received,  I've got to derive this data from  the actions not committed and the initial state.

Thank you for the feedback, I'll rework this part.

FaberVitale commented:
Applied changes based on feedback provided https://github.com/reduxjs/redux-devtools/pull/750#issuecomment-890590243

Updated demo: https://rtk-query-monitor-demo.netlify.app/
